### DISCUSSION TITLE
Some thougths after testing

### DISCUSSION THREAD

**Discussion by formeo14** [kamilkosek/jellyplist#60] [General]
First off, I want to acknowledge that the app is still in active development, likely in an alpha or beta phase. With that in mind, I’ll share some of my thoughts and observations as I’ve been testing it out.

### **Setup and Configuration**
I was able to get Jellyplist running, but there were a few points I didn’t fully understand:

1. **Database Credentials:**
   - What is the difference between `POSTGRES_USER` and `JELLYPLIST_DB_USER`, and similarly, `POSTGRES_PASSWORD` and `JELLYPLIST_DB_PASSWORD`? 
   - Since Postgres is the database, are these settings related to Redis instead?

2. **Secret Key:**
   - What is the purpose of the `SECRET_KEY`?

3. **Playlist Updates:**
   - Does Jellyplist automatically update playlists when they are modified on Spotify?

4. **Admin Credentials:**
   - I’m not entirely comfortable with Jellyplist handling my admin password. Why doesn’t an API key suffice for creating playlists? Have you raised an issue with Jellyfin to support this? Given the app is still in development, I’m cautious about fully trusting the login form at this stage.

5. **Database Connection String:**
   - I would have preferred a simpler one-liner for the database connection, such as:  
     `postgres://user:password@host/database?sslmode=disable`

### **User Interface and Experience**
1. **Login Page:**
   - The login page mentions "Login with Jellyfin credentials" twice. I think the red infobox is unnecessary. Instead, it would be helpful to include a link to the password reset page, similar to Jellyseerr.

2. **Home Page:**
   - The homepage doesn’t render well on mobile devices. Elements are misplaced, which is understandable given the app is still in development.
   - I suggest consolidating the search functionality into a single search bar that can parse both playlist strings and links. This would streamline the process since playlists follow a specific format.
   - The house icon (`\f015`) doesn’t display on Firefox for Android.

3. **Spotify Integration:**
   - It would be great if users could log in and authenticate with Spotify directly. This is crucial because dynamically generated playlists often yield different results unless authenticated.
   - Additionally, it would be helpful to display the user’s saved playlists as an option, especially for those who don’t have a Spotify account. In the future, support for other platforms like Deezer, Tidal, or Qobuz could be considered.

4. **Lidarr Integration:**
   - I encountered an **Internal Server Error** on the Lidarr page:  
     `exceptions.HTTPError: 401 Client Error: Unauthorized for url: http://lidarr:9696/api/v1/qualityprofile`.  
     I double-checked the URI and API key, and they seem correct.
   - I recommend allowing users to input their Lidarr credentials directly in the settings rather than relying on environment variables.

5. **Playlist Management:**
   - When I added a playlist, it downloaded the songs without my notice into a `__jellyplist` directory, bypassing Lidarr.  
     - **Why is this happening?**  
     - **Can this behavior be disabled?**  
     - **Is there a way to send download requests to Lidarr instead?**

6. **Jellyfin Integration:**
   - Jellyfin is referenced locally in Docker with a hostname like `http://Jellyfin:8096`. However, on my local network, it has an IP address like `http://192.168.0.100:8096`, and externally, it’s accessible via `https://jellyfin.mydomain.com`.  
     - The "Play with Jellyfin" button doesn’t work because no device recognizes `Jellyfin:8096`.  
     - **Can we have separate internal and external URLs for Jellyfin?**

7. **Miscellaneous UI Issues:**
   - When opening Spotify track information, the overlay sometimes doesn’t close properly, leaving a gray layer on the screen.
   - The play button sometimes displays inconsistent icons, such as `▶▶`, `▶▶▶`, `▶▶⏹`, or `▶⏹`.    <img align="right" src="https://github.com/user-attachments/assets/3d664b5a-8a5a-4a8e-87e8-ff49b33b5d21" alt="logo" height="100" />
   - It would be nice to indicate who created the playlist (e.g., Spotify or another user).
   - The "Last updated" date on the playlist is not formatted correctly, for example: `Last Updated: 2024-12-20 08:25:00.006414 | Last Change: None`

8. **Access Control:**
   - There’s currently no way to restrict access to song downloads for specific users or set a limit on requests/downloads per user.

### **Priorities and Final Thoughts**
While there are several areas for improvement, some features seem more critical than others:
- **Spotify authentication**
- **External URL support for Jellyfin**
- **API key support instead of username/password**

Despite the issues, Jellyplist shows a lot of promise and has a solid foundation. I’m excited to see how it evolves!
Thank you for developing it! It's truly fantastic that someone took on this missing feature.

**Comment by kamilkosek** [comment]
Hi, 
thanks for this detailed review. 
My key takeaway is: I need to work on the documentation 👍 

> Database Credentials:

You are right, it is redundant and can be unified.

> Secret Key

[https://flask.palletsprojects.com/en/stable/config/#SECRET_KEY](https://flask.palletsprojects.com/en/stable/config/#SECRET_KEY)

> Playlist Updates

Yes, the playlists are updated regularly: tracks are added, removed and moved in order to mirror the playlist from the provider. 

> Admin Credentials 

See here: [https://github.com/jellyfin/jellyfin/discussions/12868](https://github.com/jellyfin/jellyfin/discussions/12868) The Admin credentials are needed to be able to add and remove users (share playlists) to playlists. 
Basically the workflow like this: 
1. Normal User logs to Jellyplist and adds a playlist to his account.
2. If the Playlist is already requested by any other user, Jellyplist takes the admin credentials and adds the _normal User_ to the playlist.
3. If the Playlist **isn´t** already requested, Jellyplist takes the admin credentials, creates the playlist and adds the _normal user_ to the playlist. 

So the whole "admin credentials" story, is just to be able to maintain only one copy of a playlist while being able to share it accross multiple users. 

> Database Connection String

This topic is currently wip . In the process of unifing the container´s , this will probably be the endresult, so that. i.e. `SQLite` will also be supported. 


>## User Interface and Experience

I know, the UI and UX are pretty poor, this is due to some design decisions I´ve made in the process of creating this project. It all emerged from a simple python script. 
I think the best would be a complete rework of the UI, with `vue` or something similar. But first priority currently is implementing additional providers (deezer, ytm and so on) , deemix in addition to spotDL and possibly different media backend support.


> Login Page:

This is due to the fact when you try to access `/`, if you are not authenticated you will be redirected to the login page, and the error message at the bottom flashes up.

> Home Page:
The house icon (\f015) doesn’t display on Firefox for Android.

The mobile view is something I haven´t had time to work on. It´s basically the default bootstrap5 behaviour. 

> Spotify Integration:
> 
> It would be great if users could log in and authenticate with Spotify directly. This is crucial because dynamically generated playlists often yield different results unless authenticated.
> Additionally, it would be helpful to display the user’s saved playlists as an option, especially for those who don’t have a Spotify account. In the future, support for other platforms like Deezer, Tidal, or Qobuz could be considered.

This is currently impossible, see here: [https://github.com/kamilkosek/jellyplist/issues/25](https://github.com/kamilkosek/jellyplist/issues/25)

> Lidarr Integration:
> 
> I encountered an Internal Server Error on the Lidarr page:
> exceptions.HTTPError: 401 Client Error: Unauthorized for url: http://lidarr:9696/api/v1/qualityprofile.
> I double-checked the URI and API key, and they seem correct.

Strange, but I will take a look at it. 

> I recommend allowing users to input their Lidarr credentials directly in the settings rather than relying on environment variables.
Moving some settings which are not crucial to be set at startup, will be moved to the UI little by little.


> Playlist Management:
> When I added a playlist, it downloaded the songs without my notice into a __jellyplist directory, bypassing Lidarr.
> Why is this happening?

Jellyplist uses `spotDL` under the hood, it cant be disabled right now, but will be in the future. But you can change the target folder. 
The initial idea behind this is: When you add a playlist, I´d like to have available asap, so `spotDL` takes care of the tracks from the playlist if they aren´t found already in Jellyfin, and downloads them. Later if Lidarr or other tools, provide the same track but with a better quality, Jellyplist will replace it in the playlist.

> Jellyfin Integration:

> Jellyfin is referenced locally in Docker with a hostname like http://Jellyfin:8096. However, on my local network, it has an IP address like http://192.168.0.100:8096, and externally, it’s accessible via https://jellyfin.mydomain.com.
> The "Play with Jellyfin" button doesn’t work because no device recognizes Jellyfin:8096.
> Can we have separate internal and external URLs for Jellyfin?

Good catch! Will take care of it. 

> Miscellaneous UI Issues:
> 
> When opening Spotify track information, the overlay sometimes doesn’t close properly, leaving a gray layer on the screen.
> The play button sometimes displays inconsistent icons, such as ▶▶, ▶▶▶, ▶▶⏹, or ▶⏹.logo

as mentioned earlier, the UI needs a lot of refactoring. 

> It would be nice to indicate who created the playlist (e.g., Spotify or another user).

So the owner of the playlist at the provider level, or at Jellyplist level ? 

> The "Last updated" date on the playlist is not formatted correctly, for example: Last Updated: 2024-12-20 08:25:00.006414 | Last Change: None

This is fixed in the next release.

> Access Control:
> 
> There’s currently no way to restrict access to song downloads for specific users or set a limit on requests/downloads per user.

This could be easily implemented. 



Again, thanks for the time taking for testing this out and writing this review.

**Reply by formeo14** [reply]
>Secret Key

Could the secret key not be generated under the hood, or is it necessary that the user/admin configure it?

> This is due to the fact when you try to access `/`, if you are not authenticated, you will be redirected to the login page, and the error message at the bottom flashes up.

My reverse proxy does not let me remove the `/` at `sub.mydomain.xxxx/` so this shows up always. Can you configure it to not show on the `/` path?

> The mobile view is something I haven’t had time to work on. It’s basically the default Bootstrap 5 behavior.

It’s okay. Desktop is enough for now.

> This is currently impossible, see here: https://github.com/kamilkosek/jellyplist/issues/25

No, I wouldn’t say that. It is possible, as authenticating a user provides the playlists the user created. By adding the session cookie of the user, you also get the Spotify-generated playlists for a user. So a user could also get their own playlists after the change in the Spotify API.

> Jellyplist uses spotDL under the hood, it can’t be disabled right now, but will be in the future.

Great! Sorry I cannot help—I have no experience with Python GUI. I’ve only heard from friends that Python is really not good for GUIs or is really complex and slow. Don’t know if that’s true, but good luck!

> So the owner of the playlist at the provider level, or at Jellyplist level?

The Spotify owner.

> This could be easily implemented.

Thank you! I really like this app, even though it is still in early development. It works—I ask for a playlist, and it loads it. It would be cool to have user login, but it is still great. When there is a Lidarr Spotify -> YouTube downloader plugin or Soulseek plugin, I think this will be much better. And when Jellyfin includes PostgreSQL as a database soon instead of SQLite. Now there is a Meilisearch plugin for Jellyfin, so searches are also faster, and your app really completes the setup for music management! But I cannot say how it integrates into Lidarr as it does not work as of now.
